biohansel/create/cluster_generator.py

```diff
@@ -0,0 +1,236 @@
+        parent_hc_clusters = output_subtypes[sorted_thresholds[threshold_index - 1]]
+
+        threshold = sorted_thresholds[threshold_index]
+        cluster_genomes = clusters_genomes_dict[threshold]
+
+        for subtype, parent_genomes in parent_hc_clusters.items():
+            if len(parent_genomes) < min_group_size:
+                continue
+            subclade = 1
+            for _, child_genomes in cluster_genomes.items():
+                if len(child_genomes) < min_group_size:
+                    continue
+
+                if parent_genomes == child_genomes:
+                    output_subtypes[threshold][subtype] = parent_genomes
+                elif child_genomes.issubset(parent_genomes):
+                    new_subgroup_name = f'{subtype}.{subclade}'
+                    if len(parent_genomes - child_genomes) >= min_group_size:
+                        output_subtypes[threshold][new_subgroup_name] = child_genomes
+                    else:
+                        output_subtypes[threshold][subtype] = child_genomes
+                    subclade += 1
+
+    return output_subtypes
+
+
+def expand_sets(cluster_dict: Dict[str, Dict[str, Set[str]]]) -> Dict[str, Dict[str, str]]:
+    """
+    Fills in the rest of the cells in the dataframe that had previously been unfilled
+
+    Args:
+        cluster_dict: a dictionary that contains the cluster groupings of each genome with each value being sets of
+                    genomes
+
+    Returns:
+        modified_cluster_dict: A dictionary with values beings filled for all cells
+
+    """
+
+    modified_cluster_dict = {}
+    for threshold, groupings in cluster_dict.items():
+        threshold_dict = {}
+        for grouping, genomes in groupings.items():
+            for genome in genomes:
+                threshold_dict[genome] = grouping
+        modified_cluster_dict[threshold] = threshold_dict
+
+    return modified_cluster_dict
+
+
+def row_subtype(curr_genome: pd.Series) -> str:
+    """
+    Provides the last valid subtype for that particular genome, as that's the last valid branch point
+
+    Args:
+        curr_genome: the current subtypes for that particular genome
+
+    Returns:
+        row_unique[-1]: the last valid subtype name for that genome
+
+    """
+
+    row_unique = curr_genome.unique()
+    row_unique = row_unique[(row_unique != '') & (~pd.isnull(row_unique))]
+
+    return row_unique[-1]
+
+
+def df_to_subtypes_dict(cluster_genomes_df: pd.DataFrame) -> Dict[str, str]:
+    """
+    Provides a dictionary of the subtypes with the genome being the key and the subtype being the value
+
+    Args:
+        cluster_genomes_df: the dataframe containing the genomes and their respective subtypes at each threshold
+
+    Returns:
+        final_cluster_dict: the dictionary that contains the final subtype assignments for each genome
+
+    """
+
+    final_cluster_dict = {}
+    for genome, row in cluster_genomes_df.apply(row_subtype, axis=1).iteritems():
```

Comment:
[`pd.Series.to_dict()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.to_dict.html)?

Reply:
I kept this as a function because the values would then get passed to row_subtype after